### ISSUE TITLE
[bug?] cython_reset when using ranges in azimuthalIntegrator

### ISSUE THREAD

**Issue by malte-storm** [silx-kit/pyFAI#1579]
Hi,

I tested the integration with radial_range keyword and noticed that is it much slower than without. After some digging, I think found the reason for this.
If the azimuth_range or radial_range keywords are used, the integration engine compares the stored ranges with the new ones. For some reason, the CSR / LUT is re-calculated for every integration because the checks fail. As an example (from the azimuthalIntegrator.integrate1d_ng method): `cython_integr.pos0Range != (min(radial_range), max(radial_range) * EPS32)`

I do not understand why the upper range is multiplied by the EPS32 value. I ran some tests (using python 3.9 and pyFAI 0.20.0) on Windows 10 and Debian 11 and the given input range and the stored cython_integr.pos0Range were always identical. I tested this using both float32 and float64 ranges and this made no difference for me.

As an example, running the following code (incomplete, for demonstration but I can supply the full code if needed):
``` 
_method = 'csr_ocl'
_unit = 'q_nm^-1'
_rad_range = (6.63176365145901, 19.844819228621045)

ai = AzimuthalIntegrator()
ai.read(_ponif)
with Timer():
    q, I = ai.integrate1d(raw, 1600, method=_method, unit=_unit)
with Timer():
    q, I = ai.integrate1d(raw, 1600, method=_method, unit=_unit)

_N = np.where((q >= _rad_range[0]) & (q <= _rad_range[1]))[0].size

ai = AzimuthalIntegrator()
ai.read(_ponif)
with Timer():
    q, I = ai.integrate1d(raw, _N, method=_method, unit=_unit,
                      radial_range=_rad_range)
with Timer():
    q, I = ai.integrate1d(raw, _N, method=_method, unit=_unit,
                          radial_range=_rad_range)

from pyFAI.method_registry import IntegrationMethod
from pyFAI.utils import EPS32
method = ai._normalize_method(_method, dim=1, default=ai.DEFAULT_METHOD_1D)
cython_method = IntegrationMethod.select_method(
    method.dimension, method.split_lower, method.algo_lower, "cython")[0]
cython_engine = ai.engines[cython_method]
cython_integr = cython_engine.engine

pos0_scale = pyFAI.units.to_unit(_unit).scale
radial_range = tuple(_rad_range[i] / pos0_scale for i in (0, -1))

print('Input range: ', _rad_range)
print('Converted range: ', radial_range)
print('Cython integrator range: ', cython_integr.pos0Range)
print('Check range: ', (min(radial_range), max(radial_range) * EPS32))

```
yielded the following results:
```
    Code running in 3.264638 seconds.
    Code running in 0.010999 seconds.
    Code running in 2.412179 seconds.
    Code running in 0.383326 seconds.
    Input range:  (6.63176365145901, 19.844819228621045)
    Converted range:  (6.63176365145901, 19.844819228621045)
    Cython integrator range:  (6.63176365145901, 19.844819228621045)
    Check range:  (6.63176365145901, 19.844821594307845)
```

While I understand the long runtime for the first call after the new instantiation of the AzimuthalIntegrator, I would expect the second call to be comparable in runtime for both (or even slightly faster for the range). As it is, the setup_CSR is called again which increases the runtime considerably.

I do not understand the reason for multiplying the upper limit of the ranges by the EPS32 value. If I change the code of the AzimuthalIntegrator and remove the EPS32 multiplication, I get the following runtimes (using csr_ocl):
```
Code running in 2.284377 seconds.
Code running in 0.010939 seconds.
Code running in 2.133763 seconds.
Code running in 0.009711 seconds.
Input range:  (6.63176365145901, 19.844819228621045)
Converted range:  (6.63176365145901, 19.844819228621045)
Cython integrator range:  (6.63176365145901, 19.844819228621045)
modified Check range:  (6.63176365145901, 19.844819228621045)
```
These are more consistent and what I would expect. Is the EPS32 legacy code needed for an earlier Python version / pyFAI version?

**Comment by kif** [comment]
Concerning the bug, it reminds me of this one:
https://github.com/silx-kit/pyFAI/issues/1510
which was fixed, back in May (thus pyFAI version 0.20 is still affected by this bug).

Concerning the upper-bound multiplied by EPS32: EPS32 is the smallest floating point value greater than 1.0 one can represent in IEEE754 (32 bits).
For histograming, if the bin goes from `min` to `max` value, then the `min` value will fall into the bin and `max` will fall into the next bin. To prevent this behaviour (and have the `max` value also falling into the bin), pyFAI increases slightly the upper-limit. This is why there is `min`, `maxin` (maximum value inside the bin) and `max` which is maxin*EPS32.

I use EPS32 (and not EPS64) because on GPU, when float64 is not available, the numerical value of EPS64 becomes 1 and the problem arrives again.

**Comment by malte-storm** [comment]
Okay, thanks for the explanation with the histogram. 

Looking at the latest version of the azimuthalIntegration in master, the multiplication with EPS32 has been removed (except for the integrate1d_legacy, integrate_2d_legacy and sigma_clip_ng). So this is probably the similar bug as reported (and fixed) in #1510. Sorry for raising the issue again.

**Comment by kif** [comment]
Well, the code has probably moved following a refactoring ... but the idea remains the same.
There will be huge changes for the next release of pyFAI with all cython code basically rewritten.
